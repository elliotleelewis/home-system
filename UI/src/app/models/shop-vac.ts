export interface ShopVac {
	id: string;
	isOpen: boolean;
	createdAt: Date;
	updatedAt: Date;
}
