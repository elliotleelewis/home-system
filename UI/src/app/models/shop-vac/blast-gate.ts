export interface BlastGate {
	id: string;
	isOpen: boolean;
	createdAt: Date;
	updatedAt: Date;
}
