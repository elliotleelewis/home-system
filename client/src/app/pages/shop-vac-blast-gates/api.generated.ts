/* eslint-disable @typescript-eslint/array-type,@typescript-eslint/naming-convention */
import * as Types from '../../../../../node_modules/@app/schema';

export type BlastGatesQueryVariables = Types.Exact<{ [key: string]: never }>;

export type BlastGatesQuery = {
	__typename?: 'Query';
	blastGates: Array<{
		__typename?: 'BlastGate';
		id: string;
		name: string;
		isOpen: boolean;
	}>;
};

/* eslint-enable */
