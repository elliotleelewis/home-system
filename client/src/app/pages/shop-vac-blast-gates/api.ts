import { gql } from 'apollo-angular';

export const BLAST_GATES = gql`
	query BlastGates {
		blastGates {
			id
			name
			isOpen
		}
	}
`;
