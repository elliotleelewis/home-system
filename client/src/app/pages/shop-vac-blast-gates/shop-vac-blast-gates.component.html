<div class="card flex flex-col">
	<div class="rounded-t-2xl bg-base-300 p-2">
		<div class="flex items-center justify-between">
			<h2 class="p-2 text-lg font-bold">Blast Gates</h2>
			<div>
				<div class="flex gap-2">
					<button class="btn" (click)="openAllBlastGates()">
						Open All
					</button>
					<button class="btn" (click)="closeAllBlastGates()">
						Close All
					</button>
					<button
						class="btn"
						[ngClass]="{
							'btn-success': !create,
							'btn-error': create,
						}"
						(click)="toggleCreate()"
					>
						<span class="text-white">
							<i class="fas fa-plus" *ngIf="!create"></i>
							<i class="fas fa-ban" *ngIf="create"></i>
						</span>
					</button>
				</div>
			</div>
		</div>
	</div>
	<app-blast-gate
		class="card-body bg-base-200"
		[ngClass]="{
			'rounded-b-2xl': !create,
		}"
		*ngFor="let blastGate of blastGates; trackBy: trackByBlastGatesId"
		[blastGate]="blastGate"
		(updateBlastGate)="updateBlastGate($event)"
		(deleteBlastGate)="deleteBlastGate($event)"
		(activateBlastGate)="activateBlastGate($event)"
	></app-blast-gate>
	<form
		class="flex justify-between rounded-b-2xl bg-base-300 p-2"
		*ngIf="create"
		(ngSubmit)="createBlastGate()"
	>
		<div class="flex items-center gap-2">
			<input
				class="toggle"
				type="checkbox"
				name="is-open"
				[(ngModel)]="createModel.blastGateInput.isOpen"
			/>
			<div class="join">
				<input
					class="input join-item"
					type="text"
					name="name"
					[(ngModel)]="createModel.blastGateInput.name"
				/>
				<button class="btn btn-success join-item" type="submit">
					Create
				</button>
			</div>
		</div>
		<button class="btn btn-warning" type="button" disabled>Activate</button>
	</form>
</div>
