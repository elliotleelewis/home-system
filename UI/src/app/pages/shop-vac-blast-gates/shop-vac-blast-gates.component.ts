import { Component, OnDestroy, OnInit, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';

import { BlastGate } from '../../models/shop-vac/blast-gate';
import { BlastGateCreate } from '../../models/shop-vac/dto/blast-gate-create';
import { ShopVacHubConnectionService } from '../../services/shop-vac-hub-connection.service';
import { ShopVacService } from '../../services/shop-vac.service';

@Component({
	selector: 'app-shop-vac-blast-gates',
	templateUrl: './shop-vac-blast-gates.component.html',
	styleUrls: ['./shop-vac-blast-gates.component.scss'],
})
export class ShopVacBlastGatesComponent implements OnInit, OnDestroy {
	@ViewChild('createForm', { static: false })
	createForm: NgForm;
	create = false;
	createModel: BlastGateCreate = {
		id: '',
		isOpen: false,
	};

	private _blastGates: BlastGate[] = [];

	constructor(
		private _shopVac: ShopVacService,
		private _shopVacUpdate: ShopVacHubConnectionService,
	) {}

	get blastGates(): BlastGate[] {
		return this._blastGates;
	}

	set blastGates(blastGates: BlastGate[]) {
		this._blastGates = blastGates
			.map((blastGate) => {
				blastGate.createdAt = new Date(String(blastGate.createdAt));
				blastGate.updatedAt = new Date(String(blastGate.updatedAt));
				return blastGate;
			})
			.sort((a, b) => a.createdAt.valueOf() - b.createdAt.valueOf());
	}

	ngOnInit(): void {
		this._shopVacUpdate.start();
		this._shopVac.getAllBlastGates().subscribe((blastGates) => {
			this.blastGates = blastGates;
			this._shopVacUpdate.blastGates.subscribe(
				(updatedBlastGates) => (this.blastGates = updatedBlastGates),
			);
		});
	}

	ngOnDestroy(): void {
		this._shopVacUpdate.stop();
	}

	toggleCreate(): void {
		this.create = !this.create;
	}

	createBlastGate(): void {
		this._shopVac.createBlastGate(this.createModel).subscribe();
		this.create = false;
		this.createModel = {
			id: '',
			isOpen: false,
		};
	}

	updateBlastGate(blastGate: BlastGate, event: Event): void {
		this._shopVac
			.updateBlastGate(blastGate, {
				isOpen: (event.target as HTMLInputElement).checked,
			})
			.subscribe();
	}

	activateBlastGate(blastGate: BlastGate): void {
		this._shopVac.activateBlastGate(blastGate).subscribe();
	}

	openAllBlastGates(): void {
		this._shopVac.openAllBlastGates().subscribe();
	}

	closeAllBlastGates(): void {
		this._shopVac.closeAllBlastGates().subscribe();
	}
}
