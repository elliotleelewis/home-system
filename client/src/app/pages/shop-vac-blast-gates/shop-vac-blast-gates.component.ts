import { NgClass } from '@angular/common';
import {
	ChangeDetectionStrategy,
	ChangeDetectorRef,
	Component,
	type OnDestroy,
	type OnInit,
	inject,
} from '@angular/core';
import { FormsModule } from '@angular/forms';
import { type BlastGate, type MutationUpsertBlastGateArgs } from '@app/schema';
import { SubSink } from 'subsink';

import { ShopVacService } from '../../services/shop-vac.service';

import { BlastGateComponent } from './components/blast-gate/blast-gate.component';

@Component({
	selector: 'app-shop-vac-blast-gates',
	templateUrl: './shop-vac-blast-gates.component.html',
	changeDetection: ChangeDetectionStrategy.OnPush,
	imports: [NgClass, BlastGateComponent, FormsModule],
})
export class ShopVacBlastGatesComponent implements OnInit, OnDestroy {
	private readonly _changeDetectorRef = inject(ChangeDetectorRef);
	private readonly _shopVacService = inject(ShopVacService);

	create = false;
	createModel: MutationUpsertBlastGateArgs = {
		blastGateInput: {
			id: null,
			name: '',
			isOpen: false,
		},
	};

	blastGates: BlastGate[] = [];

	private _subs = new SubSink();

	ngOnInit(): void {
		this._subs.sink = this._shopVacService
			.getBlastGates()
			.subscribe((blastGates) => {
				this.blastGates = blastGates;
				this._changeDetectorRef.detectChanges();
			});
	}

	ngOnDestroy(): void {
		this._subs.unsubscribe();
	}

	toggleCreate(): void {
		this.create = !this.create;
	}

	createBlastGate(): void {
		this._subs.sink = this._shopVacService
			.upsertBlastGate(this.createModel)
			.subscribe();
		this.create = false;
		this.createModel = {
			blastGateInput: {
				id: null,
				name: '',
				isOpen: false,
			},
		};
	}

	updateBlastGate({ id, name, isOpen }: BlastGate): void {
		this._subs.sink = this._shopVacService
			.upsertBlastGate({
				blastGateInput: {
					id,
					name,
					isOpen,
				},
			})
			.subscribe();
	}

	deleteBlastGate(blastGate: BlastGate): void {
		this._subs.sink = this._shopVacService
			.deleteBlastGate({ blastGateId: blastGate.id })
			.subscribe();
	}

	activateBlastGate(blastGate: BlastGate): void {
		this._subs.sink = this._shopVacService
			.activateBlastGate({ blastGateId: blastGate.id })
			.subscribe();
	}

	openAllBlastGates(): void {
		this._subs.sink = this._shopVacService.openAllBlastGates().subscribe();
	}

	closeAllBlastGates(): void {
		this._subs.sink = this._shopVacService.closeAllBlastGates().subscribe();
	}
}
