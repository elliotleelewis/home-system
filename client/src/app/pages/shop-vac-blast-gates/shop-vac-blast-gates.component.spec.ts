import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { of } from 'rxjs';

import { ShopVacService } from '../../services/shop-vac.service';

import { ShopVacBlastGatesComponent } from './shop-vac-blast-gates.component';

describe('ShopVacBlastGatesComponent', () => {
	let component: ShopVacBlastGatesComponent;
	let fixture: ComponentFixture<ShopVacBlastGatesComponent>;

	let mockShopVacService: Partial<ShopVacService>;

	beforeEach(
		waitForAsync(() => {
			mockShopVacService = {
				getBlastGates: jest.fn(() =>
					of([
						{
							id: '12345',
							name: 'Test',
							isOpen: true,
						},
					]),
				),
			};

			void TestBed.configureTestingModule({
				declarations: [
					ShopVacBlastGatesComponent,
				],
				imports: [FormsModule],
				providers: [
					{ provide: ShopVacService, useValue: mockShopVacService },
				],
			}).compileComponents();
		}),
	);

	beforeEach(() => {
		fixture = TestBed.createComponent(ShopVacBlastGatesComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should render', () => {
		expect(fixture).toMatchSnapshot();
	});
});
