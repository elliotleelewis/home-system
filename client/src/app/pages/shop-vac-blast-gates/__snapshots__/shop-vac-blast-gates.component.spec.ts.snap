// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopVacBlastGatesComponent should render 1`] = `
<app-shop-vac-blast-gates
  _changeDetectorRef={[Function ViewRef$1]}
  _shopVacService={[Function Object]}
  _subs={[Function SubSink]}
  blastGates={[Function Array]}
  create="false"
  createModel={[Function Object]}
>
  <div
    class="card flex flex-col"
  >
    <div
      class="rounded-t-2xl bg-base-300 p-2"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="p-2 text-lg font-bold"
        >
          Blast Gates
        </h2>
        <div>
          <div
            class="flex gap-2"
          >
            <button
              class="btn"
            >
               Open All 
            </button>
            <button
              class="btn"
            >
               Close All 
            </button>
            <button
              class="btn btn-success"
            >
              <span
                class="text-white"
              >
                <i
                  class="fas fa-plus fa-fw"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body bg-base-200 rounded-b-2xl"
    >
      <app-blast-gate />
    </div>
  </div>
</app-shop-vac-blast-gates>
`;
